kf_lib_data_ingest/etl/extract/utils.py

```diff
@@ -0,0 +1,292 @@
+                )
+                col_index = list(col_series.index) * round(length_multiplier)
+                if index:
+                    if index != col_index:
+                        raise Exception(
+                            "Inconsistent column indices.", index, col_index
+                        )
+                else:
+                    index = col_index
+        df_out.index = index
+        return df_out, skip_messages
+
+    def extract(self, df, extract_cfg_or_path, apply_after_read_func=True):
+        """
+        Apply the operations in an extract config to the DataFrame to
+        produce a clean DataFrame with columns mapped to the standard
+        set of columns defined in the concept schema:
+        kf_lib_data_ingest.common.concept_schema
+
+        :param df: source data
+        :type df: pandas.DataFrame
+        :param extract_cfg_or_path: either the ExtractConfig object or path to the
+        extract config so the ExtractConfig object can be instantiated
+        :type extract_cfg_or_path: str or ExtractConfig
+
+        :returns: the extracted pandas.DataFrame
+        """
+        # Check inputs
+        assert_safe_type(df, pandas.DataFrame)
+        assert_safe_type(extract_cfg_or_path, str, ExtractConfig)
+        if df.empty:
+            raise Exception("Extraction failed! DataFrame cannot be empty")
+
+        # Load extract config
+        if isinstance(extract_cfg_or_path, str):
+            self.extract_config = ExtractConfig(extract_cfg_or_path)
+        else:
+            self.extract_config = extract_cfg_or_path
+            self.extract_config._validate()
+
+        extract_config = self.extract_config
+
+        # Clean df
+        df_in = clean_up_df(df)
+
+        # Optionally post-process df
+        if apply_after_read_func and extract_config.do_after_read:
+            self.logger.info("Calling custom do_after_read function.")
+            df_in = extract_config.do_after_read(df)
+
+            # Check for empty df
+            if len(df_in.index) == 0:
+                raise ConfigValidationError(
+                    "Source DataFrame is empty. Check your do_after_read function"
+                )
+
+        # Describe Dataframe
+        self.logger.debug(f"Read DataFrame with dimensions {df_in.shape}")
+        self.logger.debug(f"Column headers are: {list(df_in.columns)}")
+
+        # Check operations
+        if not extract_config.operations:
+            self.logger.info("The operation list is empty. Nothing to do.")
+            return None
+
+        # Run extraction
+        df_out, file_skips = self._chain_operations(
+            df_in, extract_config.operations
+        )
+
+        # record skipped operations
+        if file_skips:
+            self.messages.append(
+                f"In file {extract_config.config_file_relpath}:"
+            )
+            self.messages.extend(f"\t{s}" for s in file_skips)
```

Comment:
It looks like now if you use the same Extract instance multiple times the internal messages list will get longer and longer. That appears to not be intentional since the ExtractStage uses the same instance over and over and extends the outer message list each time. Changing this to be a local messages list here instead of self.messages and returning it to the caller at the end of this method and extending the outer message list with the returned list would align more closely with the previous operation.

Reply:
I can't return messages since the extract function needs to return the extracted df. Instead I just reset `self.messages = []` at the top of extract and then extract stage can access it the way it does now.

Reply:
You can return more than one thing.

Like...previously ExtractStage._run() would call _chain_operations many times, once for each extract configuration. And now ExtractStage._run() will call Extractor.extract many times, once for each extract configuration, in pretty much exactly the same way.

So if Extractor.extract is just a thin wrapper around _chain_operations, then it could return the same way that _chain_operations does.

But anyway this works too. I just think it looks a little funny to have an instance variable inside of Extractor for something that doesn't preserve state across invocations. But then I think that making Extractor a class at all instead of just a module function is kinda weird too. Not weird enough to not approve the change though.